VideoLocker/src/main/java/org/edx/mobile/http/OauthRefreshTokenAuthenticator.java

```diff
@@ -53,14 +56,30 @@ public OauthRefreshTokenAuthenticator(Context context) {
     public Request authenticate(Route route, final Response response) throws IOException {
         logger.warn(response.toString());
 
-        if (!isTokenExpired(response.peekBody(HttpStatus.OK).string())) {
-            return null;
-        }
-
         final AuthResponse currentAuth = loginPrefs.getCurrentAuth();
         if (null == currentAuth || null == currentAuth.refresh_token) {
             return null;
         }
+        String response_body = response.peekBody(200).string();
+
+        switch (getErrorCode(response_body)) {
+            case TOKEN_EXPIRED_ERROR_MESSAGE:
+                break;
```

Comment:
Also, you're not returning `null` in the `TOKEN_NONEXISTENT_ERROR_MESSAGE` and `TOKEN_INVALID_GRANT_ERROR_MESSAGE` case if the header is not found. Instead it's falling through to the `TOKEN_EXPIRED_ERROR_MESSAGE` case, which breaks.

What I had in mind was something like this:

```` java
String error_code = getErrorCode(response_body);
if (error_code != null) {
    switch (error_code) {
        case TOKEN_EXPIRED_ERROR_MESSAGE: {
            final AuthResponse refreshedAuth;
            try {
                refreshedAuth = refreshAccessToken(currentAuth);
            } catch (HttpResponseStatusException e) {
                break;
            }
            return response.request().newBuilder()
                    .header("Authorization", refreshedAuth.token_type + " " + refreshedAuth.access_token)
                    .build();
        }
        case TOKEN_NONEXISTENT_ERROR_MESSAGE:
        case TOKEN_INVALID_GRANT_ERROR_MESSAGE: {
            // Retry request with the current access_token if the original access_token used in
            // request does not match the current access_token. This case can occur when
            // asynchronous calls are made and are attempting to refresh the access_token where
            // one call succeeds but the other fails. https://github.com/edx/edx-app-android/pull/834
            if (!response.request().headers().get("Authorization").split(" ")[1].equals(currentAuth.access_token)) {
                return response.request().newBuilder()
                        .header("Authorization", currentAuth.token_type + " " + currentAuth.access_token)
                        .build();
            }
        }
    }
}
return null;
````

This seems like the most simple flow to me. But it's up to you.